### DISCUSSION TITLE
M600 Macro and KAMP

### DISCUSSION THREAD

**Discussion by Grogamanda1** [xblax/flashforge_ad5m_klipper_mod#91] [General]
Hi All, 

HAs anyone got an M600 Macro for the AD5M or is it the same as the Stock Klipper Macro? Need the printer to pause and home when slicer calls M600 and then I can resume after LOAD_FILAMENT.

Also - does KAMP work on this mod?

I only installed the mod on my AD5M today so havent done very much with it.

Cheers in advance

**Comment by xblax** [comment]
I did now implement a M600 macro. Will be in the next release:
https://github.com/xblax/flashforge_adm5_klipper_mod/commit/db0f0fc8fec5c92a88935b73b29d3c38732bdf19

![image](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/9802544/5d6056d1-ac51-4f7f-8455-4e9e8366328c)

Regarding KAMP, I guess it would work if somebody writes a new installer script or does a manual install. But it's not very useful for printers with nozzle based probe. For that reason I don't consider to add it to the mod by default. Cleaning the nozzle via Heating, Wipe, Cooldown cycle takes longer than the actual leveling. It makes more sense to do a full mesh occasionally then adaptive mesh before each print.

**Reply by DougJoseph** [reply]
> I did now implement a M600 macro. Will be in the next release:

I'm guessing that, in order for this to work, then I need to, in OrcaSlicer, for my "Machine G-code" options for the AD5M, in the box labelled "Change Filament G-code" I need to input "M600" and save changes, correct? I say this because without that, I searched for M600 in the Gcode outputted with a filament change requested, and it seems it is missing from the outputted file.

**Reply by DougJoseph** [reply]
Also of interest, when I did the above (in OrcaSlicer, for my "Machine G-code" options for the AD5M, in the box labelled "Change Filament G-code" I inputted "M600" and saved changes), then OrcaSlicer outputs M600 **_twice_**:  once at the beginning of the first layer, and a second time at the layer I specified to have a filament change be invoked. So, the job paused needlessly at the start, until I pressed to continue.

**Reply by DougJoseph** [reply]
Update: so it's not working for me to get a filament change during a print by putting "M600" into the Machine G-code for the "Change Filament Box" — at least when I deleted the unneeded first M600, and left the needed second M600, the printer seems to interpret that M600 as "job complete" and so while it does stop printing, it does not reckon the job to be paused but rather concluded. I attempted putting "resume" into the console, but I just got a response saying the printer was not at a pause state, and thus no resume. So, I replaced M600 in OrcaSlicer in the Machine G-Code with M25, and I'm going to see how that works.

**Reply by DougJoseph** [reply]
Another update: inputting M25 as the filament change code in OrcaSlicer in the Machine G-Code... also did not work correctly. As before it drops two M25's in the outputted code, one at the first layer and a second one at the specified layer, and the first one "works" as intended, but the second causes the printer to think the job is complete instead of doing a pause state. Interestingly, if I edit the Gcode file and delete the first M25, the second one (which becomes the only one) still does not work. I will need to do more testing, including on my other printer as well, to see how reproducible this is.

**Comment by greenman360** [comment]
KAMP works for me.  With a caveat.  I installed it using the guide on the GitHub, through SSH.  I then changed/added all the various settings and code in the config files and my slicer.

Adaptive Mesh works, as does Smart Parking.  It'll mesh the area the model is in and then park the printhead to the bottom left of it.  But no matter what I've tried (and I've spent over a day now tearing out my hair), I can't get the Adaptive Purge to work consistently, or basically at all.  It'll do the purge, whether set to standard or the fancy Voron purge, but it almost always does it dead center of the bed.  Randomly, for reasons I can't figure out at all, it will print the purge both how and where it should in relation to the model's placement.  But even using the same settings without changing any configuration, another sliced model will put it right in the middle again.

**Reply by Raymond812** [reply]
I got the AD5M too. I'm having trouble to setup the KAMP properly.
Would you mind share the information how to set it up.
I able to get the [exclude_object] load up but not the Adaptive mesh and the purge
I updated the printer.cfg and the moonraker.cfg maybe the path is kind of mess up.
I'm using orca slicer. I updated the Starting gcode to
START_PRINT BED_TEMP={first_layer_bed_temperature[0]} EXTRUDER_TEMP={first_layer_temperature[0]}

Thanks in advance
